### PR TITLE
[move] Refactored dynamic fields natives work with Move Prover

### PR DESCRIPTION
**Pull request by awelc**
This refactoring is needed in order for the Prover to correctly process a native function whose return value is a mutable reference that cannot be shown to come from any of the parameter values. In particular, the Prover compiler complains about the following definition:
```
public(friend) native fun borrow_child_object<Child: key>(parent: address, id: address): &mut Child;
```

but is OK with the following definition:
```
public(friend) native fun borrow_child_object_mut<Child: key>(object: &mut UID, id: address): &mut Child;
```

One question here is whether this refactoring (passing &[mut] UID` instead of `address` should be applied to all native functions for consistency.

This PR is predicated on the acceptance of the https://github.com/move-language/move/pull/679 core Move PR and will not land before the core Move PR is accepted in some shape or form.

### CODE REVIEW

File: crates/sui-framework/sources/dynamic_field.move [hunk 1]
```diff
@@ -120,28 +120,43 @@ public fun exists_with_type<Name: copy + drop + store, Value: store>(
...
 
 public(friend) native fun add_child_object<Child: key>(parent: address, child: Child);
 
 /// throws `EFieldDoesNotExist` if a child does not exist with that ID
 /// or throws `EFieldTypeMismatch` if the type does not match
-public(friend) native fun borrow_child_object<Child: key>(parent: address, id: address): &mut Child;
+/// we need two versions to return a reference or a mutable reference
+public(friend) native fun borrow_child_object<Child: key>(object: &UID, id: address): &Child;
+public(friend) native fun borrow_child_object_mut<Child: key>(object: &mut UID, id: address): &mut Child;
+
```
Comment (tnowacki) [inline crates/sui-framework/sources/dynamic_field.move:159]: nit remove extra space

File: crates/sui-framework/src/natives/dynamic_field.rs [hunk 2]
```diff
@@ -146,7 +151,15 @@ pub fn borrow_child_object(
     assert!(ty_args.len() == 1);
     assert!(args.len() == 2);
     let child_id = pop_arg!(args, AccountAddress).into();
-    let parent = pop_arg!(args, AccountAddress).into();
+
+    let parent_uid = pop_arg!(args, StructRef).read_ref().unwrap();
+    // UID { id: ID { bytes: address } }
```
Comment (tnowacki) [inline crates/sui-framework/src/natives/dynamic_field.rs:156]: 👍

**Review by tnowacki — COMMENTED** [review]
Looks great! I'll hold off approving until we find out if we need it for the prover?

**Comment by awelc** [comment]
> Looks great! I'll hold off approving until we find out if we need it for the prover?

The core Move part of this has been approved (https://github.com/move-language/move/pull/679) and changes in this Sui PR verified to work with those in the core Move part. I have corrected the extra space issue so hopefully this should be good to land.

**Pull request merged**